libvips/foreign/tiff2vips.c

```diff
@@ -1056,6 +1058,76 @@ rtiff_onebit_line( Rtiff *rtiff, VipsPel *q, VipsPel *p, int n, void *flg )
 	}
 }
 
+/* Per-scanline process function for 2 bit greyscale images.
+ */
+static void
+rtiff_twobit_line( Rtiff *rtiff, VipsPel *q, VipsPel *p, int n, void *flg )
+{
+        int x, i, z;
+        int minisblack = 
+			rtiff->header.photometric_interpretation == PHOTOMETRIC_MINISBLACK;
+        VipsPel bits;
+
+        x = 0;
+        for( i = 0; i < (n >> 2); i++ ) {
+                bits = (VipsPel) minisblack ? p[i] : ~p[i];
+
+                for( z = 0; z < 4; z++ ) {
+					/* The grey shade is the value four times concatenated */
+                        q[x] = (bits & 0xC0) | ((bits & 0xC0) >> 2) 
+							| ((bits & 0xC0) >> 4) | (bits >> 6);
```

Comment:
You could write this as:

```C
                        VipsPel fourbits = (bits & 0xC0) | ((bits & 0xC0) >> 2);
                        q[x] = fourbits | (fourbits >> 4);
```

Though I've no idea if that gives a useful speedup.

Reply:
Yes one could do that for sure and I also think it is easier to understand. I'll change it in the next iteration of my proposal. Any further ideas?

Reply:
Oh I guess you could remove the `&` as well if you build from the low bits up:

```C
                        VipsPel twobits = bits >> 6;
                        VipsPel fourbits = twobits | (twobits << 2);
                        VipsPel eightbits = fourbits | (fourbits << 4);

                        q[x] = eightbits;
```